File: ios/Throttler.swift
import Foundation

protocol NetworkQualityScheduledTask: AnyObject {
  func cancel()
}

extension DispatchWorkItem: NetworkQualityScheduledTask {}

final class Throttler {
  typealias Clock = () -> Double
  typealias Scheduler = (
    _ delayMs: Double,
    _ action: @escaping () -> Void
  ) -> NetworkQualityScheduledTask
  typealias Emitter = ([String: Any]) -> Void

  private static let numericKeys = [
    "downlinkKbps",
    "uplinkKbps",
    "signalStrength",
  ]

  private var throttleMs: Double
  private var bandwidthChangeThresholdPct: Double
  private let clock: Clock
  private let scheduler: Scheduler
  private let emit: Emitter

  private var lastEmitted: [String: Any]?
  private var lastEmitTime = -Double.infinity
  private var pending: [String: Any]?
  private var timer: NetworkQualityScheduledTask?

  init(
    throttleMs: Double,
    bandwidthChangeThresholdPct: Double,
    clock: @escaping Clock,
    scheduler: @escaping Scheduler,
    emit: @escaping Emitter
  ) {
    self.throttleMs = max(0, throttleMs)
    self.bandwidthChangeThresholdPct = max(0, bandwidthChangeThresholdPct)
    self.clock = clock
    self.scheduler = scheduler
    self.emit = emit
  }

  func update(throttleMs: Double, bandwidthChangeThresholdPct: Double) {
    self.throttleMs = max(0, throttleMs)
    self.bandwidthChangeThresholdPct = max(0, bandwidthChangeThresholdPct)
    guard let pending else { return }
    cancelPending()
    submit(pending)
  }

  func submit(_ snapshot: [String: Any]) {
    guard let previous = lastEmitted else {
      emitNow(snapshot)
      return
    }

    if Self.equalsIgnoringTimestamp(previous, snapshot) {
      cancelPending()
      return
    }

    if Self.isSignificantChange(previous, snapshot) {
      cancelPending()
      emitNow(snapshot)
      return
    }

    let changedNumericKeys = Self.numericKeys.filter {
      !Self.valuesEqual(previous[$0], snapshot[$0])
    }
    let crossesThreshold = changedNumericKeys.contains {
      Self.exceedsThreshold(
        oldValue: previous[$0],
        newValue: snapshot[$0],
        thresholdPct: bandwidthChangeThresholdPct
      )
    }
    guard crossesThreshold else {
      cancelPending()
      return
    }

    let now = clock()
    let elapsed = now - lastEmitTime
    if elapsed >= throttleMs {
      cancelPending()
      emitNow(snapshot)
      return
    }

    pending = snapshot
    guard timer == nil else { return }
    timer = scheduler(max(0, throttleMs - elapsed)) { [weak self] in
      guard let self else { return }
      self.timer = nil
      guard let pending = self.pending else { return }
      self.pending = nil
      self.emitNow(pending)
    }
  }

  func reset() {
    cancelPending()
    lastEmitted = nil
    lastEmitTime = -Double.infinity
  }

  static func equalsIgnoringTimestamp(
    _ first: [String: Any],
    _ second: [String: Any]
  ) -> Bool {
    PathSnapshot.comparableKeys.allSatisfy {
      valuesEqual(first[$0], second[$0])
    }
  }

  static func isSignificantChange(
    _ first: [String: Any],
    _ second: [String: Any]
  ) -> Bool {
    PathSnapshot.comparableKeys
      .filter { !numericKeys.contains($0) }
      .contains { !valuesEqual(first[$0], second[$0]) }
  }

  static func exceedsThreshold(
    oldValue: Any?,
    newValue: Any?,
    thresholdPct: Double
  ) -> Bool {
    let oldNumber = numericValue(oldValue)
    let newNumber = numericValue(newValue)
    if oldNumber == nil || newNumber == nil {
      return oldNumber != nil || newNumber != nil
    }
    guard let oldNumber, let newNumber else { return false }
    if oldNumber == newNumber { return false }
    if oldNumber == 0 { return true }
    return abs(newNumber - oldNumber) / abs(oldNumber) * 100 >= thresholdPct
  }

  private func emitNow(_ snapshot: [String: Any]) {
    lastEmitted = snapshot
    lastEmitTime = clock()
    emit(snapshot)
  }

  private func cancelPending() {
    timer?.cancel()
    timer = nil
    pending = nil
  }

  private static func numericValue(_ value: Any?) -> Double? {
    guard !(value is NSNull) else { return nil }
    return (value as? NSNumber)?.doubleValue
  }

  private static func valuesEqual(_ first: Any?, _ second: Any?) -> Bool {
    switch (first, second) {
    case (nil, nil):
      return true
    case let (first as NSObject, second as NSObject):
      return first.isEqual(second)
    default:
      return false
    }
  }
}

File: android/src/main/java/com/rnnetworkquality/Throttler.kt
package com.rnnetworkquality

internal fun interface Cancellable {
  fun cancel()
}

internal fun interface Clock {
  fun nowMs(): Long
}

internal fun interface Scheduler {
  fun schedule(delayMs: Long, task: () -> Unit): Cancellable
}

/**
 * De-duplicates snapshots and applies leading/trailing throttling to minor signal changes.
 * All calls must be serialized by the owner; the class itself has no platform dependencies.
 */
internal class Throttler(
  private val clock: Clock,
  private val scheduler: Scheduler,
  private val emit: (NetworkSnapshot) -> Unit,
  throttleMs: Long,
  bandwidthChangeThresholdPct: Double,
) {
  private var throttleMs = throttleMs.coerceAtLeast(0L)
  private var bandwidthChangeThresholdPct = bandwidthChangeThresholdPct.coerceAtLeast(0.0)
  private var lastEmitted: NetworkSnapshot? = null
  private var lastEmitTime = 0L
  private var pending: NetworkSnapshot? = null
  private var timer: Cancellable? = null

  fun updateOptions(throttleMs: Long, bandwidthChangeThresholdPct: Double) {
    this.throttleMs = throttleMs.coerceAtLeast(0L)
    this.bandwidthChangeThresholdPct = bandwidthChangeThresholdPct.coerceAtLeast(0.0)

    val latestPending = pending ?: return
    cancelPending()
    submit(latestPending)
  }

  fun submit(snapshot: NetworkSnapshot) {
    val previous = lastEmitted
    if (previous == null) {
      emitNow(snapshot)
      return
    }

    if (snapshot.equalsIgnoringTimestamp(previous)) {
      cancelPending()
      return
    }

    if (Mappers.isSignificantChange(previous, snapshot)) {
      cancelPending()
      emitNow(snapshot)
      return
    }

    if (!Mappers.exceedsAnyMinorThreshold(previous, snapshot, bandwidthChangeThresholdPct)) {
      cancelPending()
      return
    }

    val elapsed = (clock.nowMs() - lastEmitTime).coerceAtLeast(0L)
    if (elapsed >= throttleMs) {
      cancelPending()
      emitNow(snapshot)
      return
    }

    pending = snapshot
    if (timer == null) {
      timer = scheduler.schedule(throttleMs - elapsed) {
        timer = null
        val trailing = pending
        pending = null
        if (trailing != null) emitNow(trailing)
      }
    }
  }

  fun reset() {
    cancelPending()
    lastEmitted = null
    lastEmitTime = 0L
  }

  private fun emitNow(snapshot: NetworkSnapshot) {
    lastEmitted = snapshot
    lastEmitTime = clock.nowMs()
    emit(snapshot)
  }

  private fun cancelPending() {
    timer?.cancel()
    timer = null
    pending = null
  }
}
